### DISCUSSION TITLE
How to increase IO efficiency when using python embedding

### DISCUSSION THREAD

**Discussion by sdrabenh** [dtcenter/METplus#1447] [File I/O] [answered]
Are there any ways or examples showing how python embedding can be used more efficiently for large data sets? For example, say I have a generic python embedding script "fetch_fcst_array.py UGRD 20220201_0000 P500" that operates on netcdf files which do not contain standard names/conventions. In this case, it opens the file, locates the correct dimension indices, extracts out the appropriate 2D array from a 4D data set, and fills out the appropriate "attrs" object required to run grid_stat. This works fine, but now I want to apply this same method for 30 levels, 120 time steps, and 10 different variables. It happens to be the case that all this data is contained within the same large file. However, multiple calls of this embedded script creates a lot of unnecessary overhead from continually opening and closing the same file, repeatedly reading and searching through the same coordinate information, and returning the appropriate 2D slices. Tools like xarray make it much easier and faster to retrieve these arrays (even if using multiple files in a time series). Is there a way to directly pass a "met_data" array and "attrs" dictionary (currently in memory) to grid_stat from within a running python script in order to mitigate redundancy and overhead while increasing efficiency? If you have any examples of this it would be greatly appreciated.

**Comment by CPKalb** [comment, marked as answer]
Hi:

Currently there isn't an option to directly pass a met_Data array to grid_stat from within a running python script, although the idea has been brought up.  One option that has been used is to save the met_data in something like .npy format and just restore this array in the python script that MET calls.

**Reply by sdrabenh** [reply]
Thanks for the suggestion. This is another idea that could work. However, it still involves 2 disk operations to save and load. I'm unsure if it would save much overhead, but in theory, it might be faster indexing a smaller file. I would have to time it. I was hoping there was a solution that would allow these arrays to be passed in memory.

**Reply by CPKalb** [reply]
Unfortunately there isn't a way to directly pass a python array to MET in memory.   MET requires a script to run to go grab data.  MET tools are not directly callable via Python, so setting this up would involve a python module to wrap calls to the MET executables, which doesn't currently exist

**Comment by CPKalb** [comment]
Hi:

Do you have additional questions on this?  If not, I would like to close the discussion.

Thanks,
Christina
